services:
  db:
    image: 'postgres:latest'
    container_name: postgres_sqlalchemy
    env_file:
      - .env
    ports:
      - "5432:5432"
    volumes:
      - ./init.sql:/docker-entrypoint-initdb.d/init.sql
      - postgres_sqlalchemy_data:/var/lib/postgresql/data/
    networks:
      - fastapi_network
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -h 127.0.0.1"]
      interval: 10s
      timeout: 5s
      retries: 5
      start_period: 20s

  pgadmin:
    image: dpage/pgadmin4
    container_name: pgadmin_sqlalchemy
    ports:
      - "3333:80"
    env_file:
      - .env
    depends_on:
      db:
        condition: service_healthy
    volumes:
      - pgadmin_sqlalchemy_data:/var/lib/pgadmin
    networks:
      - fastapi_network

  web:
    restart: always
    build: .
    container_name: backend_fastapi
    command: ["/bin/bash", "/commands/run_web_server_dev.sh"]
    environment:
      - PYTHONUNBUFFERED=True
      - PYTHONPATH=/usr/src/fastapi
      - WATCHFILES_FORCE_POLLING=true
    env_file:
      - .env
    ports:
      - "8000:8000"
    volumes:
      - ./src:/usr/src/fastapi
    depends_on:
      db:
        condition: service_healthy
    networks:
      - fastapi_network

  migrator:
    build: .
    container_name: alembic_migrator
    command: ["/bin/bash", "/commands/run_migration.sh"]
    depends_on:
      db:
        condition: service_healthy
    volumes:
      - ./src:/usr/src/fastapi
    env_file:
      - .env
    environment:
      - PYTHONPATH=/usr/src/fastapi
    networks:
      - fastapi_network


volumes:
  postgres_sqlalchemy_data:
    driver: local
  pgadmin_sqlalchemy_data:
    driver: local

networks:
  fastapi_network:
    driver: bridge
